### DISCUSSION TITLE
KML node clamped to terrain

### DISCUSSION THREAD

**Discussion by marcoma9023** [pelicanmapping/osgearth#2461] [Support Q&A] [answered]
Hi,

I have OsgEarth 2.10 and I have a KML file relative to a geographic area.

The kml has altitude = 0 and clampToGround property.

Is there the possibility into OsgEarth to clamp the kml to the terrain in case of enabling/disabling a DEM?

Thanks

**Comment by gwaldron** [comment]
The KML clamping options are in `KML_Geometry.cpp` around line 186. This is in the master branch, but you can look there in 2.x to find the equivalent code and see if you can get it to do what you want. Good luck.

**Comment by marcoma9023** [comment]
@gwaldron Thanks for reply,

so are you suggesting that, if in the KML xml I specify the "clampToGround" option, this should automatically adapt to the terrain and I don't have to do development on the OsgEarth side?

I currently integrate it like this:
`osgEarth::Drivers::KMLOptions kmlOption;
		kmlOption.declutter() = false;
kmlHighNode = osgEarth::Drivers::KML::load(osgEarth::URI(kmlUrl), mapNode, kmlOption);
				kmlHighNode->getOrCreateStateSet()->setMode(GL_DEPTH_TEST, true);
				kmlHighNode->getOrCreateStateSet()->setDefine(OE_LIGHTING_DEFINE, osg::StateAttribute::OFF | osg::StateAttribute::PROTECTED);

mapNode->addChild(kmlHighNode );
`

where in the KML file the altitude mode is already clampToGraound, but without effect.

**Reply by gwaldron** [reply, marked as answer]
If I recall correctly, only KML lines and points will clamp to the terrain dynamically. Anything that's extruded will not.

**Comment by bclay1297** [comment]
Were you able to get KML files to load?  I have this code but nothing shows
